src/crimson/os/seastore/journal/circular_bounded_journal.h

```diff
@@ -0,0 +1,286 @@
+  >;
+  using mkfs_ret = mkfs_ertr::future<>;
+  mkfs_ret mkfs(mkfs_config_t& config);
+
+
+  /**
+   * CBJournal structure
+   *
+   * +-------------------------------------------------------+
+   * |   header    | record | record | record | record | ... |
+   * +-------------------------------------------------------+
+   *               ^-----------block aligned-----------------^
+   * <----fixed---->
+   */
+
+
+  /**
+   *
+   * CBJournal write
+   *
+   * NVMe will support a large block write (< 512KB) with atomic write unit command.
+   * With this command, we expect that the most of incoming data can be stored
+   * as a single write call, which has lower overhead than existing
+   * way that uses a combination of system calls such as write() and sync().
+   *
+   */
+
+  struct cbj_header_t {
+    uint64_t magic;
+    uuid_d uuid;
+    uint64_t block_size; // aligned with block_size
+    uint64_t size;   // max length of journal
+    uint64_t used_size;  // current used_size of journal
+    uint32_t error;      // reserved
+
+    rbm_abs_addr start_offset;      // start offset of CBJournal
+    rbm_abs_addr last_committed_record_base;
+    rbm_abs_addr written_to;
+    rbm_abs_addr applied_to;
+
+    uint64_t flag;       // represent features (reserved)
+    uint8_t csum_type;   // type of checksum algoritghm used in cbj_header_t
+    uint64_t csum;       // checksum of entire cbj_header_t
+    uint32_t cur_segment_seq;
+
+    rbm_abs_addr start; // start address of the device
+    rbm_abs_addr end;   // start address of the device
+    device_id_t device_id;
+
+    DENC(cbj_header_t, v, p) {
+      DENC_START(1, 1, p);
+      denc(v.magic, p);
+      denc(v.uuid, p);
+      denc(v.block_size, p);
+      denc(v.size, p);
+      denc(v.used_size, p);
+      denc(v.error, p);
+
+      denc(v.start_offset, p);
+
+      denc(v.last_committed_record_base, p);
+      denc(v.written_to, p);
+      denc(v.applied_to, p);
+
+      denc(v.flag, p);
+      denc(v.csum_type, p);
+      denc(v.csum, p);
+      denc(v.cur_segment_seq, p);
+      denc(v.start, p);
+      denc(v.end, p);
+      denc(v.device_id, p);
+
+      DENC_FINISH(p);
+    }
+  };
+
+  /**
+   *
+   * Write position for CBJournal
+   *
+   * | written to rbm |    written length to CBJournal    | new write |
+   * ----------------->----------------------------------->------------>
+   *                  ^      	            ^             ^
+   *            applied_to   last_committed_record_base   written_to
+   *
+   */
+
+  size_t get_used_size() const {
+    return header.used_size;
+  }
+  void set_used_size(size_t size) {
+    header.used_size = size;
+  }
+  size_t get_total_size() const {
+    return header.size;
+  }
+  rbm_abs_addr get_start_addr() const {
+    return header.start_offset;
+  }
+  size_t get_available_size() const {
+    return get_total_size() - get_used_size();
+  }
+
+  void update_applied_to(rbm_abs_addr addr, uint32_t len) {
```

Comment:
This doesn't appear to have any callers.

Reply:
This is used by unittest-seastore-cbjournal for a test purpose because there is no cleaner implementation for now. I explained applied_to as below.